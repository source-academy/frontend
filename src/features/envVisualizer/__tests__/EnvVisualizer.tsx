import { runInContext } from 'js-slang/dist/';
import createContext from 'js-slang/dist/createContext';
import { Config } from 'src/features/envVisualizer/EnvVisualizerConfig';

import { AgendaItemComponent } from '../compactComponents/AgendaItemComponent';
import { StashItemComponent } from '../compactComponents/StashItemComponent';
import { ArrayUnit } from '../components/ArrayUnit';
import { ArrowFromArrayUnit } from '../components/arrows/ArrowFromArrayUnit';
import { Frame } from '../components/Frame';
import { ArrayValue } from '../components/values/ArrayValue';
import { FnValue } from '../components/values/FnValue';
import { GlobalFnValue } from '../components/values/GlobalFnValue';
import EnvVisualizer from '../EnvVisualizer';
import { Layout } from '../EnvVisualizerLayout';
import { Env, EnvTree } from '../EnvVisualizerTypes';

// The following are code samples that are more complex/known to have caused bugs
// Some are commented out to keep the tests shorter
// Old code samples: https://github.com/source-academy/frontend/blob/006fcd47abea62f704c9d5901bd8e80097122705/src/features/envVisualizer/__tests__/EnvVisualizer.tsx
const codeSamples = [
  `
    const fn = () => "L";
    const x = ["long string", pair(() => 1, () => 2), list(1, pair(2, 3), () => 3), () => "THIS", 5, 6];
    const y = list(x[1], x[2], tail(x[1]), tail(x[2]), fn);        
    debugger;`,
  `
    const fn = () => 1;
    const x = [1, pair(pair(1, 2), 3), 4];
    const l = list(1, list(list(1, 2, 3, fn),x[1] , [1, pair(1,2), fn, 4, 5], 4), () => 9, x);
    debugger;`,
  `
    const x1 = list(1, 2);
    const x2 = list(3, 4);
    set_head(tail(x2), x1);
    set_tail(tail(x1), x2);
    debugger;`,
  `
    const e = list(null, 2, list(3,4,5));
    set_head(e, head(tail(tail(e))));
    const f = pair(1,2);
    const g = list(null, 2, list(() => 1,4,5));
    set_head(g, head(tail(tail(g))));
    const y = pair(null, pair(() => 2, pair(1,2)));
    set_head(y, tail(tail(y)));
    debugger;`,
  `
    let x = 0;
    let y = 10;
    function f(x) {
        return () => 6;
    }
    const z = f(11);
    debugger;`,
  `
    const x = list(1,pair, accumulate);
    debugger;`,
  `
    const x = [];
    debugger;`,
  `let y = [];
    for (let i = 1; i < 5; i = i + 1) {
        const x = [y];
    }
    y = [1, [1, 2, 3]];
    y[1][1] = y;
    y[0] = y;
    y[1][2] = y[1];
    function eval_stream(s, n) {
        function es(s, n) {
            const z = [y];
            return n === 1 ? list(head(s)) : pair(head(s), es(stream_tail(s), n - 1));
        }
        return n === 0 ? null : es(s, n);
    }
    eval_stream(integers_from(1), 2);
    let x = 1;
    debugger;
    `
];

codeSamples.forEach((code, idx) => {
  test('EnvVisualizer calculates correct layout for code sample ' + idx, async () => {
    if (EnvVisualizer.getCompactLayout()) {
      EnvVisualizer.toggleCompactLayout();
    }
    const context = createContext(4);
    await runInContext(code, context);
    Layout.setContext(
      context.runtime.environmentTree as EnvTree,
      context.runtime.agenda!,
      context.runtime.stash!
    );

    const toTest: any[] = [];
    const environmentsToTest: Env[] = [];
    Layout.compactLevels.forEach(({ frames }) => {
      frames.forEach(({ environment, bindings }) => {
        environmentsToTest.push(environment);
        bindings.forEach(({ keyString, data }) => {
          toTest.push(keyString);
          toTest.push(data);
        });
      });
    });
    Layout.grid.frameLevels.forEach(({ frames }) => {
      frames.forEach(({ environment, bindings, xCoord, yCoord }) => {
        environmentsToTest.push(environment);
        toTest.push(xCoord);
        toTest.push(yCoord);
        bindings.forEach(({ keyString, data }) => {
          toTest.push(keyString);
          toTest.push(data);
        });
      });
    });
    environmentsToTest.forEach(environment => {
      expect(environment).toMatchSnapshot({
        id: expect.any(String)
      });
    });
    expect(toTest).toMatchSnapshot();
    const checkNonCompactLayout = () => {
      Layout.draw();
      Layout.values.forEach(v => {
        if (v instanceof GlobalFnValue || v instanceof FnValue) {
          const arrow = v.arrow();
          expect(arrow).toBeDefined();
          expect(arrow?.target).toBeDefined();
          const path = arrow!.path().match(/[^ ]+/g) ?? [];
          expect(path.length).toEqual(14);
          expect(path[1]).toEqual(path[4]); // move up
          expect(path[8]).toEqual(path[10]); // move left
          expect(Frame.lastXCoordBelow(v.x())).toEqual(Frame.lastXCoordBelow(arrow!.target!.x())); // target
        } else if (v instanceof ArrayValue) {
          v.arrows().forEach(arrow => {
            expect(arrow).toBeDefined();
            expect(arrow?.target).toBeDefined();
            if (
              arrow instanceof ArrowFromArrayUnit &&
              arrow.target instanceof ArrayValue &&
              arrow.source instanceof ArrayUnit
            ) {
              const sourceArray = arrow.source.parent as ArrayValue;
              const targetArray = arrow.target as ArrayValue;
              if (sourceArray.level === targetArray.level) {
                // for arrows within same array level
                const path = arrow!.path().match(/[^ ]+/g) ?? [];
                expect(parseFloat(path[1])).toEqual(arrow.source.x() + Config.DataUnitWidth / 2);
                expect(parseFloat(path[2])).toEqual(arrow.source.y() + Config.DataUnitHeight / 2);
                if (sourceArray.data === targetArray.data) {
                  expect(path.length).toEqual(22); // up, right, down.
                  expect(path[1]).toEqual(path[4]);
                  expect(path[17]).toEqual(path[20]);
                  expect(parseFloat(path[20]) - parseFloat(path[1])).toBeCloseTo(
                    Config.DataUnitWidth / 3
                  );
                }
              }
            }
          });
        }
      });
    };
    const checkCompactLayout = () => {
      Layout.draw();
    };
    checkCompactLayout();
    EnvVisualizer.togglePrintableMode();
    checkCompactLayout();
    EnvVisualizer.toggleCompactLayout();
    checkNonCompactLayout();
    EnvVisualizer.togglePrintableMode();
    checkNonCompactLayout();
  });
});

const codeSamplesAgendaStash = [
  [
    'arrows from the environment instruction to the frame and arrows from the stash to closures',
    `
    function create(n) {
      const arr = [];
      let x = 0;
      
      while (x < n) {
          arr[x] = () => x;
          x = x + 1;
      }
      return arr;
    }
    create(3)[1]();
    `,
    33
  ],
  [
    'global environments are treated correctly',
    `
    {
      const math_sin = x => x;
    }
    math_sin(math_PI / 2); 
    `,
    7
  ],
  [
    'Agenda is truncated properly',
    `
      function fact(n) {
        return n <= 1 ? 1 : n * fact(n - 1);
      }
      fact(10);
      `,
    161,
    true
  ]
];

codeSamplesAgendaStash.forEach((codeSample, idx) => {
  test('EnvVisualizer Agenda Stash correctly renders: ' + codeSample[0], async () => {
    const code = codeSample[1] as string;
    const envSteps = codeSample[2] as number;
    const truncate = codeSample[3];
    if (!EnvVisualizer.getCompactLayout()) {
      EnvVisualizer.toggleCompactLayout();
    }
    if (truncate) {
      EnvVisualizer.toggleStackTruncated();
    }
    EnvVisualizer.toggleAgendaStash();
    const context = createContext(4);
    await runInContext(code, context, { executionMethod: 'ec-evaluator', envSteps: envSteps });
    Layout.setContext(
      context.runtime.environmentTree as EnvTree,
      context.runtime.agenda!,
      context.runtime.stash!
    );
    Layout.draw();
    const agendaItemsToTest: AgendaItemComponent[] = Layout.agendaComponent.stackItemComponents;
    const stashItemsToTest: StashItemComponent[] = Layout.stashComponent.stashItemComponents;
    agendaItemsToTest.forEach(item => {
      expect(item.draw()).toMatchSnapshot();
      if (item.value === 'ENVIRONMENT') expect(item.arrow).toBeDefined();
    });
    if (truncate) expect(agendaItemsToTest.length).toBeLessThanOrEqual(10);
    stashItemsToTest.forEach(item => {
      expect(item.draw()).toMatchSnapshot();
    });
  });
});
