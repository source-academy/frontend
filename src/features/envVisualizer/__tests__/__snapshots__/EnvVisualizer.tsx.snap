// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EnvVisualizer calculates correct layout for code sample 0 1`] = `
Object {
  "head": Object {
    ":::pre-declared names::": Symbol(),
  },
  "id": Any<String>,
  "name": "global",
  "tail": null,
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 0 2`] = `
Object {
  "head": Object {
    "3": [Function],
    "4": [Function],
    "5": [Function],
    "6": [Function],
    "7": [Function],
    "8": [Function],
    "9": [Function],
    "fn": [Function],
    "x": Array [
      "long string",
      Array [
        [Function],
        [Function],
      ],
      Array [
        1,
        Array [
          Array [
            2,
            3,
          ],
          Array [
            [Function],
            null,
          ],
        ],
      ],
      [Function],
      5,
      6,
    ],
    "y": Array [
      Array [
        [Function],
        [Function],
      ],
      Array [
        Array [
          1,
          Array [
            Array [
              2,
              3,
            ],
            Array [
              [Function],
              null,
            ],
          ],
        ],
        Array [
          [Function],
          Array [
            Array [
              Array [
                2,
                3,
              ],
              Array [
                [Function],
                null,
              ],
            ],
            Array [
              [Function],
              null,
            ],
          ],
        ],
      ],
    ],
  },
  "id": Any<String>,
  "name": "programEnvironment",
  "tail": Object {
    "head": Object {
      "$accumulate": [Function],
      "$append": [Function],
      "$build_list": [Function],
      "$enum_list": [Function],
      "$filter": [Function],
      "$length": [Function],
      "$list_to_string": [Function],
      "$map": [Function],
      "$remove": [Function],
      "$remove_all": [Function],
      "$reverse": [Function],
      "accumulate": [Function],
      "append": [Function],
      "build_list": [Function],
      "build_stream": [Function],
      "enum_list": [Function],
      "enum_stream": [Function],
      "equal": [Function],
      "eval_stream": [Function],
      "filter": [Function],
      "for_each": [Function],
      "integers_from": [Function],
      "is_list": [Function],
      "is_stream": [Function],
      "length": [Function],
      "list_ref": [Function],
      "list_to_stream": [Function],
      "list_to_string": [Function],
      "map": [Function],
      "member": [Function],
      "remove": [Function],
      "remove_all": [Function],
      "reverse": [Function],
      "stream_append": [Function],
      "stream_filter": [Function],
      "stream_for_each": [Function],
      "stream_length": [Function],
      "stream_map": [Function],
      "stream_member": [Function],
      "stream_ref": [Function],
      "stream_remove": [Function],
      "stream_remove_all": [Function],
      "stream_reverse": [Function],
      "stream_to_list": [Function],
    },
    "id": "1",
    "name": "programEnvironment",
    "tail": Object {
      "head": Object {
        ":::pre-declared names::": Symbol(),
      },
      "id": "-1",
      "name": "global",
      "tail": null,
    },
  },
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 0 3`] = `
Array [
  ":::pre-declared names::: ",
  Symbol(),
  "fn:= ",
  [Function],
  "x:= ",
  Array [
    "long string",
    Array [
      [Function],
      [Function],
    ],
    Array [
      1,
      Array [
        Array [
          2,
          3,
        ],
        Array [
          [Function],
          null,
        ],
      ],
    ],
    [Function],
    5,
    6,
  ],
  "y:= ",
  Array [
    Array [
      [Function],
      [Function],
    ],
    Array [
      Array [
        1,
        Array [
          Array [
            2,
            3,
          ],
          Array [
            [Function],
            null,
          ],
        ],
      ],
      Array [
        [Function],
        Array [
          Array [
            Array [
              2,
              3,
            ],
            Array [
              [Function],
              null,
            ],
          ],
          Array [
            [Function],
            null,
          ],
        ],
      ],
    ],
  ],
  "3:= ",
  [Function],
  "4:= ",
  [Function],
  "5:= ",
  [Function],
  "6:= ",
  [Function],
  "7:= ",
  [Function],
  "8:= ",
  [Function],
  "9:= ",
  [Function],
]
`;

exports[`EnvVisualizer calculates correct layout for code sample 1 1`] = `
Object {
  "head": Object {
    ":::pre-declared names::": Symbol(),
  },
  "id": Any<String>,
  "name": "global",
  "tail": null,
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 1 2`] = `
Object {
  "head": Object {
    "12": [Function],
    "13": [Function],
    "14": [Function],
    "15": [Function],
    "fn": [Function],
    "l": Array [
      1,
      Array [
        Array [
          Array [
            1,
            Array [
              2,
              Array [
                3,
                Array [
                  [Function],
                  null,
                ],
              ],
            ],
          ],
          Array [
            Array [
              Array [
                1,
                2,
              ],
              3,
            ],
            Array [
              Array [
                1,
                Array [
                  1,
                  2,
                ],
                [Function],
                4,
                5,
              ],
              Array [
                4,
                null,
              ],
            ],
          ],
        ],
        Array [
          [Function],
          Array [
            Array [
              1,
              Array [
                Array [
                  1,
                  2,
                ],
                3,
              ],
              4,
            ],
            null,
          ],
        ],
      ],
    ],
    "x": Array [
      1,
      Array [
        Array [
          1,
          2,
        ],
        3,
      ],
      4,
    ],
  },
  "id": Any<String>,
  "name": "programEnvironment",
  "tail": Object {
    "head": Object {
      "$accumulate": [Function],
      "$append": [Function],
      "$build_list": [Function],
      "$enum_list": [Function],
      "$filter": [Function],
      "$length": [Function],
      "$list_to_string": [Function],
      "$map": [Function],
      "$remove": [Function],
      "$remove_all": [Function],
      "$reverse": [Function],
      "accumulate": [Function],
      "append": [Function],
      "build_list": [Function],
      "build_stream": [Function],
      "enum_list": [Function],
      "enum_stream": [Function],
      "equal": [Function],
      "eval_stream": [Function],
      "filter": [Function],
      "for_each": [Function],
      "integers_from": [Function],
      "is_list": [Function],
      "is_stream": [Function],
      "length": [Function],
      "list_ref": [Function],
      "list_to_stream": [Function],
      "list_to_string": [Function],
      "map": [Function],
      "member": [Function],
      "remove": [Function],
      "remove_all": [Function],
      "reverse": [Function],
      "stream_append": [Function],
      "stream_filter": [Function],
      "stream_for_each": [Function],
      "stream_length": [Function],
      "stream_map": [Function],
      "stream_member": [Function],
      "stream_ref": [Function],
      "stream_remove": [Function],
      "stream_remove_all": [Function],
      "stream_reverse": [Function],
      "stream_to_list": [Function],
    },
    "id": "10",
    "name": "programEnvironment",
    "tail": Object {
      "head": Object {
        ":::pre-declared names::": Symbol(),
      },
      "id": "-1",
      "name": "global",
      "tail": null,
    },
  },
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 1 3`] = `
Array [
  ":::pre-declared names::: ",
  Symbol(),
  "fn:= ",
  [Function],
  "x:= ",
  Array [
    1,
    Array [
      Array [
        1,
        2,
      ],
      3,
    ],
    4,
  ],
  "l:= ",
  Array [
    1,
    Array [
      Array [
        Array [
          1,
          Array [
            2,
            Array [
              3,
              Array [
                [Function],
                null,
              ],
            ],
          ],
        ],
        Array [
          Array [
            Array [
              1,
              2,
            ],
            3,
          ],
          Array [
            Array [
              1,
              Array [
                1,
                2,
              ],
              [Function],
              4,
              5,
            ],
            Array [
              4,
              null,
            ],
          ],
        ],
      ],
      Array [
        [Function],
        Array [
          Array [
            1,
            Array [
              Array [
                1,
                2,
              ],
              3,
            ],
            4,
          ],
          null,
        ],
      ],
    ],
  ],
  "12:= ",
  [Function],
  "13:= ",
  [Function],
  "14:= ",
  [Function],
  "15:= ",
  [Function],
]
`;

exports[`EnvVisualizer calculates correct layout for code sample 2 1`] = `
Object {
  "head": Object {
    ":::pre-declared names::": Symbol(),
  },
  "id": Any<String>,
  "name": "global",
  "tail": null,
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 2 2`] = `
Object {
  "head": Object {
    "x1": Array [
      1,
      Array [
        2,
        Array [
          3,
          Array [
            [Circular],
            null,
          ],
        ],
      ],
    ],
    "x2": Array [
      3,
      Array [
        Array [
          1,
          Array [
            2,
            [Circular],
          ],
        ],
        null,
      ],
    ],
  },
  "id": Any<String>,
  "name": "programEnvironment",
  "tail": Object {
    "head": Object {
      "$accumulate": [Function],
      "$append": [Function],
      "$build_list": [Function],
      "$enum_list": [Function],
      "$filter": [Function],
      "$length": [Function],
      "$list_to_string": [Function],
      "$map": [Function],
      "$remove": [Function],
      "$remove_all": [Function],
      "$reverse": [Function],
      "accumulate": [Function],
      "append": [Function],
      "build_list": [Function],
      "build_stream": [Function],
      "enum_list": [Function],
      "enum_stream": [Function],
      "equal": [Function],
      "eval_stream": [Function],
      "filter": [Function],
      "for_each": [Function],
      "integers_from": [Function],
      "is_list": [Function],
      "is_stream": [Function],
      "length": [Function],
      "list_ref": [Function],
      "list_to_stream": [Function],
      "list_to_string": [Function],
      "map": [Function],
      "member": [Function],
      "remove": [Function],
      "remove_all": [Function],
      "reverse": [Function],
      "stream_append": [Function],
      "stream_filter": [Function],
      "stream_for_each": [Function],
      "stream_length": [Function],
      "stream_map": [Function],
      "stream_member": [Function],
      "stream_ref": [Function],
      "stream_remove": [Function],
      "stream_remove_all": [Function],
      "stream_reverse": [Function],
      "stream_to_list": [Function],
    },
    "id": "16",
    "name": "programEnvironment",
    "tail": Object {
      "head": Object {
        ":::pre-declared names::": Symbol(),
      },
      "id": "-1",
      "name": "global",
      "tail": null,
    },
  },
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 2 3`] = `
Array [
  ":::pre-declared names::: ",
  Symbol(),
  "x1:= ",
  Array [
    1,
    Array [
      2,
      Array [
        3,
        Array [
          [Circular],
          null,
        ],
      ],
    ],
  ],
  "x2:= ",
  Array [
    3,
    Array [
      Array [
        1,
        Array [
          2,
          [Circular],
        ],
      ],
      null,
    ],
  ],
]
`;

exports[`EnvVisualizer calculates correct layout for code sample 3 1`] = `
Object {
  "head": Object {
    ":::pre-declared names::": Symbol(),
  },
  "id": Any<String>,
  "name": "global",
  "tail": null,
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 3 2`] = `
Object {
  "head": Object {
    "20": [Function],
    "21": [Function],
    "e": Array [
      Array [
        3,
        Array [
          4,
          Array [
            5,
            null,
          ],
        ],
      ],
      Array [
        2,
        Array [
          Array [
            3,
            Array [
              4,
              Array [
                5,
                null,
              ],
            ],
          ],
          null,
        ],
      ],
    ],
    "f": Array [
      1,
      2,
    ],
    "g": Array [
      Array [
        [Function],
        Array [
          4,
          Array [
            5,
            null,
          ],
        ],
      ],
      Array [
        2,
        Array [
          Array [
            [Function],
            Array [
              4,
              Array [
                5,
                null,
              ],
            ],
          ],
          null,
        ],
      ],
    ],
    "y": Array [
      Array [
        1,
        2,
      ],
      Array [
        [Function],
        Array [
          1,
          2,
        ],
      ],
    ],
  },
  "id": Any<String>,
  "name": "programEnvironment",
  "tail": Object {
    "head": Object {
      "$accumulate": [Function],
      "$append": [Function],
      "$build_list": [Function],
      "$enum_list": [Function],
      "$filter": [Function],
      "$length": [Function],
      "$list_to_string": [Function],
      "$map": [Function],
      "$remove": [Function],
      "$remove_all": [Function],
      "$reverse": [Function],
      "accumulate": [Function],
      "append": [Function],
      "build_list": [Function],
      "build_stream": [Function],
      "enum_list": [Function],
      "enum_stream": [Function],
      "equal": [Function],
      "eval_stream": [Function],
      "filter": [Function],
      "for_each": [Function],
      "integers_from": [Function],
      "is_list": [Function],
      "is_stream": [Function],
      "length": [Function],
      "list_ref": [Function],
      "list_to_stream": [Function],
      "list_to_string": [Function],
      "map": [Function],
      "member": [Function],
      "remove": [Function],
      "remove_all": [Function],
      "reverse": [Function],
      "stream_append": [Function],
      "stream_filter": [Function],
      "stream_for_each": [Function],
      "stream_length": [Function],
      "stream_map": [Function],
      "stream_member": [Function],
      "stream_ref": [Function],
      "stream_remove": [Function],
      "stream_remove_all": [Function],
      "stream_reverse": [Function],
      "stream_to_list": [Function],
    },
    "id": "18",
    "name": "programEnvironment",
    "tail": Object {
      "head": Object {
        ":::pre-declared names::": Symbol(),
      },
      "id": "-1",
      "name": "global",
      "tail": null,
    },
  },
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 3 3`] = `
Array [
  ":::pre-declared names::: ",
  Symbol(),
  "e:= ",
  Array [
    Array [
      3,
      Array [
        4,
        Array [
          5,
          null,
        ],
      ],
    ],
    Array [
      2,
      Array [
        Array [
          3,
          Array [
            4,
            Array [
              5,
              null,
            ],
          ],
        ],
        null,
      ],
    ],
  ],
  "f:= ",
  Array [
    1,
    2,
  ],
  "g:= ",
  Array [
    Array [
      [Function],
      Array [
        4,
        Array [
          5,
          null,
        ],
      ],
    ],
    Array [
      2,
      Array [
        Array [
          [Function],
          Array [
            4,
            Array [
              5,
              null,
            ],
          ],
        ],
        null,
      ],
    ],
  ],
  "y:= ",
  Array [
    Array [
      1,
      2,
    ],
    Array [
      [Function],
      Array [
        1,
        2,
      ],
    ],
  ],
  "20:= ",
  [Function],
  "21:= ",
  [Function],
]
`;

exports[`EnvVisualizer calculates correct layout for code sample 4 1`] = `
Object {
  "head": Object {
    ":::pre-declared names::": Symbol(),
  },
  "id": Any<String>,
  "name": "global",
  "tail": null,
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 4 2`] = `
Object {
  "head": Object {
    "24": [Function],
    "28": [Function],
    "f": [Function],
    "x": 0,
    "y": 10,
    "z": [Function],
  },
  "id": Any<String>,
  "name": "programEnvironment",
  "tail": Object {
    "head": Object {
      "$accumulate": [Function],
      "$append": [Function],
      "$build_list": [Function],
      "$enum_list": [Function],
      "$filter": [Function],
      "$length": [Function],
      "$list_to_string": [Function],
      "$map": [Function],
      "$remove": [Function],
      "$remove_all": [Function],
      "$reverse": [Function],
      "accumulate": [Function],
      "append": [Function],
      "build_list": [Function],
      "build_stream": [Function],
      "enum_list": [Function],
      "enum_stream": [Function],
      "equal": [Function],
      "eval_stream": [Function],
      "filter": [Function],
      "for_each": [Function],
      "integers_from": [Function],
      "is_list": [Function],
      "is_stream": [Function],
      "length": [Function],
      "list_ref": [Function],
      "list_to_stream": [Function],
      "list_to_string": [Function],
      "map": [Function],
      "member": [Function],
      "remove": [Function],
      "remove_all": [Function],
      "reverse": [Function],
      "stream_append": [Function],
      "stream_filter": [Function],
      "stream_for_each": [Function],
      "stream_length": [Function],
      "stream_map": [Function],
      "stream_member": [Function],
      "stream_ref": [Function],
      "stream_remove": [Function],
      "stream_remove_all": [Function],
      "stream_reverse": [Function],
      "stream_to_list": [Function],
    },
    "id": "22",
    "name": "programEnvironment",
    "tail": Object {
      "head": Object {
        ":::pre-declared names::": Symbol(),
      },
      "id": "-1",
      "name": "global",
      "tail": null,
    },
  },
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 4 3`] = `
Object {
  "callExpression": Object {
    "arguments": Array [
      Object {
        "loc": undefined,
        "type": "Literal",
        "value": 11,
      },
    ],
    "callee": Node {
      "end": 97,
      "loc": SourceLocation {
        "end": Position {
          "column": 15,
          "line": 7,
        },
        "start": Position {
          "column": 14,
          "line": 7,
        },
      },
      "name": "f",
      "start": 96,
      "type": "Identifier",
    },
    "end": 101,
    "loc": SourceLocation {
      "end": Position {
        "column": 19,
        "line": 7,
      },
      "start": Position {
        "column": 14,
        "line": 7,
      },
    },
    "start": 96,
    "type": "CallExpression",
  },
  "head": Object {
    "27": [Function],
    "x": 11,
  },
  "id": Any<String>,
  "name": "f",
  "tail": Object {
    "head": Object {
      "24": [Function],
      "28": [Function],
      "f": [Function],
      "x": 0,
      "y": 10,
      "z": [Function],
    },
    "id": "23",
    "name": "programEnvironment",
    "tail": Object {
      "head": Object {
        "$accumulate": [Function],
        "$append": [Function],
        "$build_list": [Function],
        "$enum_list": [Function],
        "$filter": [Function],
        "$length": [Function],
        "$list_to_string": [Function],
        "$map": [Function],
        "$remove": [Function],
        "$remove_all": [Function],
        "$reverse": [Function],
        "accumulate": [Function],
        "append": [Function],
        "build_list": [Function],
        "build_stream": [Function],
        "enum_list": [Function],
        "enum_stream": [Function],
        "equal": [Function],
        "eval_stream": [Function],
        "filter": [Function],
        "for_each": [Function],
        "integers_from": [Function],
        "is_list": [Function],
        "is_stream": [Function],
        "length": [Function],
        "list_ref": [Function],
        "list_to_stream": [Function],
        "list_to_string": [Function],
        "map": [Function],
        "member": [Function],
        "remove": [Function],
        "remove_all": [Function],
        "reverse": [Function],
        "stream_append": [Function],
        "stream_filter": [Function],
        "stream_for_each": [Function],
        "stream_length": [Function],
        "stream_map": [Function],
        "stream_member": [Function],
        "stream_ref": [Function],
        "stream_remove": [Function],
        "stream_remove_all": [Function],
        "stream_reverse": [Function],
        "stream_to_list": [Function],
      },
      "id": "22",
      "name": "programEnvironment",
      "tail": Object {
        "head": Object {
          ":::pre-declared names::": Symbol(),
        },
        "id": "-1",
        "name": "global",
        "tail": null,
      },
    },
  },
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 4 4`] = `
Array [
  ":::pre-declared names::: ",
  Symbol(),
  "x: ",
  0,
  "y: ",
  10,
  "f:= ",
  [Function],
  "z:= ",
  [Function],
  "24:= ",
  [Function],
  "x: ",
  11,
  "27:= ",
  [Function],
]
`;

exports[`EnvVisualizer calculates correct layout for code sample 5 1`] = `
Object {
  "head": Object {
    ":::pre-declared names::": Symbol(),
    "accumulate": [Function],
    "pair": [Function],
  },
  "id": Any<String>,
  "name": "global",
  "tail": null,
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 5 2`] = `
Object {
  "head": Object {
    "31": [Function],
    "x": Array [
      1,
      Array [
        [Function],
        Array [
          [Function],
          null,
        ],
      ],
    ],
  },
  "id": Any<String>,
  "name": "programEnvironment",
  "tail": Object {
    "head": Object {
      "$accumulate": [Function],
      "$append": [Function],
      "$build_list": [Function],
      "$enum_list": [Function],
      "$filter": [Function],
      "$length": [Function],
      "$list_to_string": [Function],
      "$map": [Function],
      "$remove": [Function],
      "$remove_all": [Function],
      "$reverse": [Function],
      "accumulate": [Function],
      "append": [Function],
      "build_list": [Function],
      "build_stream": [Function],
      "enum_list": [Function],
      "enum_stream": [Function],
      "equal": [Function],
      "eval_stream": [Function],
      "filter": [Function],
      "for_each": [Function],
      "integers_from": [Function],
      "is_list": [Function],
      "is_stream": [Function],
      "length": [Function],
      "list_ref": [Function],
      "list_to_stream": [Function],
      "list_to_string": [Function],
      "map": [Function],
      "member": [Function],
      "remove": [Function],
      "remove_all": [Function],
      "reverse": [Function],
      "stream_append": [Function],
      "stream_filter": [Function],
      "stream_for_each": [Function],
      "stream_length": [Function],
      "stream_map": [Function],
      "stream_member": [Function],
      "stream_ref": [Function],
      "stream_remove": [Function],
      "stream_remove_all": [Function],
      "stream_reverse": [Function],
      "stream_to_list": [Function],
    },
    "id": "29",
    "name": "programEnvironment",
    "tail": Object {
      "head": Object {
        ":::pre-declared names::": Symbol(),
        "accumulate": [Function],
        "pair": [Function],
      },
      "id": "-1",
      "name": "global",
      "tail": null,
    },
  },
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 5 3`] = `
Array [
  ":::pre-declared names::: ",
  Symbol(),
  "accumulate: ",
  [Function],
  "pair: ",
  [Function],
  "x:= ",
  Array [
    1,
    Array [
      [Function],
      Array [
        [Function],
        null,
      ],
    ],
  ],
]
`;

exports[`EnvVisualizer calculates correct layout for code sample 6 1`] = `
Object {
  "head": Object {
    ":::pre-declared names::": Symbol(),
  },
  "id": Any<String>,
  "name": "global",
  "tail": null,
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 6 2`] = `
Object {
  "head": Object {
    "x": Array [],
  },
  "id": Any<String>,
  "name": "programEnvironment",
  "tail": Object {
    "head": Object {
      "$accumulate": [Function],
      "$append": [Function],
      "$build_list": [Function],
      "$enum_list": [Function],
      "$filter": [Function],
      "$length": [Function],
      "$list_to_string": [Function],
      "$map": [Function],
      "$remove": [Function],
      "$remove_all": [Function],
      "$reverse": [Function],
      "accumulate": [Function],
      "append": [Function],
      "build_list": [Function],
      "build_stream": [Function],
      "enum_list": [Function],
      "enum_stream": [Function],
      "equal": [Function],
      "eval_stream": [Function],
      "filter": [Function],
      "for_each": [Function],
      "integers_from": [Function],
      "is_list": [Function],
      "is_stream": [Function],
      "length": [Function],
      "list_ref": [Function],
      "list_to_stream": [Function],
      "list_to_string": [Function],
      "map": [Function],
      "member": [Function],
      "remove": [Function],
      "remove_all": [Function],
      "reverse": [Function],
      "stream_append": [Function],
      "stream_filter": [Function],
      "stream_for_each": [Function],
      "stream_length": [Function],
      "stream_map": [Function],
      "stream_member": [Function],
      "stream_ref": [Function],
      "stream_remove": [Function],
      "stream_remove_all": [Function],
      "stream_reverse": [Function],
      "stream_to_list": [Function],
    },
    "id": "32",
    "name": "programEnvironment",
    "tail": Object {
      "head": Object {
        ":::pre-declared names::": Symbol(),
      },
      "id": "-1",
      "name": "global",
      "tail": null,
    },
  },
}
`;

exports[`EnvVisualizer calculates correct layout for code sample 6 3`] = `
Array [
  ":::pre-declared names::: ",
  Symbol(),
  "x:= ",
  Array [],
]
`;
